src/ast/ast_build_ar_exp.c

```diff
@@ -0,0 +1,328 @@
+	}
+
+	// Alternatives
+	for(uint i = 0; i < alternatives; i++) {
+		const cypher_astnode_t *predicate = cypher_ast_case_get_predicate(expr, i);
+		op->op.children[offset++] = AR_EXP_FromExpression(record_map, predicate);
+		const cypher_astnode_t *value = cypher_ast_case_get_value(expr, i);
+		op->op.children[offset++] = AR_EXP_FromExpression(record_map, value);
+	}
+
+	// Default value.
+	const cypher_astnode_t *deflt = cypher_ast_case_get_default(expr);
+	if(deflt == NULL) {
+		// Default not specified, use NULL.
+		op->op.children[offset] = AR_EXP_NewConstOperandNode(SI_NullVal());
+	} else {
+		op->op.children[offset] = AR_EXP_FromExpression(record_map, deflt);
+	}
+
+	return op;
+}
+
+AR_ExpNode *AR_EXP_FromExpression(RecordMap *record_map, const cypher_astnode_t *expr) {
+	const cypher_astnode_type_t type = cypher_astnode_type(expr);
+
+	/* Function invocations */
+	if(type == CYPHER_AST_APPLY_OPERATOR || type == CYPHER_AST_APPLY_ALL_OPERATOR) {
+		return _AR_EXP_FromApplyExpression(record_map, expr);
+		/* Variables (full nodes and edges, UNWIND artifacts */
+	} else if(type == CYPHER_AST_IDENTIFIER) {
+		return _AR_EXP_FromIdentifierExpression(record_map, expr);
+		/* Entity-property pair */
+	} else if(type == CYPHER_AST_PROPERTY_OPERATOR) {
+		return _AR_EXP_FromPropertyExpression(record_map, expr);
+		/* SIValue constant types */
+	} else if(type == CYPHER_AST_INTEGER) {
+		return _AR_EXP_FromIntegerExpression(record_map, expr);
+	} else if(type == CYPHER_AST_FLOAT) {
+		return _AR_EXP_FromFloatExpression(record_map, expr);
+	} else if(type == CYPHER_AST_STRING) {
+		return _AR_EXP_FromStringExpression(record_map, expr);
+	} else if(type == CYPHER_AST_TRUE) {
+		return _AR_EXP_FromTruExpression(record_map, expr);
+	} else if(type == CYPHER_AST_FALSE) {
+		return _AR_EXP_FromFalseExpression(record_map, expr);
+	} else if(type == CYPHER_AST_NULL) {
+		return _AR_EXP_FromNullExpression(record_map, expr);
+		/* Handling for unary operators (-5, +a.val) */
+	} else if(type == CYPHER_AST_UNARY_OPERATOR) {
+		return _AR_EXP_FromUnaryOpExpression(record_map, expr);
+	} else if(type == CYPHER_AST_BINARY_OPERATOR) {
+		return _AR_EXP_FromBinaryOpExpression(record_map, expr);
+	} else if(type == CYPHER_AST_COMPARISON) {
+		return _AR_EXP_FromComparisonExpression(record_map, expr);
+	} else if(type == CYPHER_AST_CASE) {
+		return _AR_EXP_FromCaseExpression(record_map, expr);
+	} else {
+		/*
+		   Unhandled types:
+		   CYPHER_AST_COLLECTION
+		   CYPHER_AST_CASE
+		   CYPHER_AST_LABELS_OPERATOR
+		   CYPHER_AST_LIST_COMPREHENSION
+		   CYPHER_AST_MAP
+		   CYPHER_AST_MAP_PROJECTION
+		   CYPHER_AST_PARAMETER
+		   CYPHER_AST_PATTERN_COMPREHENSION
+		   CYPHER_AST_SLICE_OPERATOR
+		   CYPHER_AST_REDUCE
+		   CYPHER_AST_SUBSCRIPT_OPERATOR
+		*/
+		printf("Encountered unhandled type '%s'\n", cypher_astnode_typestr(type));
+		assert(false);
+	}
+
+	assert(false);
+	return NULL;
```

Comment:
This function and this whole file are great refactors; I like it a lot!

Reply:
Thanks, just following the convention you've placed with filter tree construction from AST